spec/spec.md

```diff
@@ -0,0 +1,601 @@
+~ A cryptographically secure, privacy-respecting, and machine-verifiable digital credential that contains an additional signature from a _holder_ who is authorized to present the credential to a relying party.
+
+[[def:Web5 SDK, Web5 SDKs]]
+~ A software development kit that provides developers with the tools and APIs necessary to build Web5-compatible applications.
+
+[[def:DID Resolver, DID Resolvers, DID Resolution]]
+~ A DID Resolver is a software service designed to perform resolution of a [[ref:DID Document]], which is the process of obtaining a [[ref:DID Document]] for a given [[ref:DID]]. See [[ref:DID-IDENTIFIER-RESOLUTION]].
+
+## Web5 Core Features
+
+The Web5 SDK provides a comprehensive set of features for building decentralized protocols and applications. The following section outline the core features required to build a conformant [[ref:Web5 SDK]].
+
+### Cryptographic Primitives
+
+Web5 supports the following cryptographic key types for use with the noted corresponding digital signature algorithms:
+
+| Key Type | Algorithm | Function |
+|----------|-----------|----------|
+| [[ref:secp256k1]] | `ES256K` [[spec:RFC8812]] | Signing and Verification |
+| [[ref:Ed25519]] | `EdDSA` [[spec:RFC8032]] | Signing and Verification |
+
+::: note
+In-memory signing using [[ref:secp256k1]] ****MUST**** produce k-deterministic low-s signatures with [ECDSA](https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm) as per [[spec:RFC6979]]. Verification ****MUST NOT**** require low-s signatures.
+:::
+
+### Key Management
+
+Web5 implementations ****MUST**** provide a consistent and extensible _public_ interface for key management, with the following minimum concrete implementations:
+
+1. Key Manager Interface
+2. In-Memory Key Manager
+3. AWS KMS
+4. Device Enclave (Mobile)
+5. Keychain (Mobile)
+
+::: todo
+Provide detailed specifications for each key management implementation, including APIs and usage guidelines.
+:::
+
+Further, the key manager interface ****MUST**** be passed as an argument to all public API methods that require key material such as:
+* DID Creation
+* Data Signing
+
+::: note
+[AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/asymmetric-key-specs.html#key-spec-ecc) does not support [[ref:Ed25519]]. At some point in the future, we can consider introducing support for key wrapping.
+:::
+
+Consumers of conformant [[ref:Web5 SDKs]] ****SHOULD**** be able to provide their own `KeyManager` impelementations, if desired.
+
+### JSON Schema
+
+[[ref:JSON-SCHEMA]] support is ****REQUIRED**** for multiple features throughout [[ref:Web5 SDKs]], including, but not limited to support with [[ref:verifiable credentials]], [VC JSON Schemas](#vc-json-schema), and for use in validating the data models this specification has defined.
+
+Conformant implementations of [[ref:Web5 SDKs]] ****MUST**** support at least [[ref:JSON-SCHEMA-DRAFT-7]] and [[ref:JSON-SCHEMA-2020-12]].
```

Comment:
Chad says JSON-SCHEMA-2020-12 is backwards compatible with JSON-SCHEMA-DRAFT-7, so just writing in JSON-SCHEMA-DRAFT-7 should be sufficient right?

Reply:
And if someone used features from JSON-SCHEMA-2020-12 then it may or may not be compatible with SCHEMA-DRAFT-7... lol so it seems like everyone will / should only use SCHEMA-DRAFT-7

Reply:
we should be using 2020-12 since it's the latest draft, which is being used as the input document to a LTS type draft. 

more info [here](https://json-schema.org/blog/posts/future-of-json-schema) and [here](https://json-schema.org/blog/posts/the-last-breaking-change) which notes that if you're using 2020-12 you'll be fine (we're not using breaking features)

it should be straightforward to convert our schemas to 2020-12, I've done so in the past using this tool https://github.com/sourcemeta/alterschema